flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java

```diff
@@ -508,8 +523,13 @@ public void setInitialState(@Nullable JobManagerTaskRestore taskRestore) {
 
 			final SlotRequestId slotRequestId = new SlotRequestId();
 
-			final CompletableFuture<LogicalSlot> logicalSlotFuture = preferredLocationsFuture
-				.thenCompose(
+			final ComponentMainThreadExecutor mainThreadExecutor =
+				vertex.getExecutionGraph().getJobMasterMainThreadExecutor();
+
+			final CompletableFuture<LogicalSlot> logicalSlotFuture =
+				FutureUtils.composeAsyncIfNotDone(
+					preferredLocationsFuture,
+					mainThreadExecutor,
```

Comment:
I think that is a good idea to prevent some of the scattered `async(..., mainThread)` calls where we could instead just proceed if we already sync back right here.

Reply:
Yes, it looks like the only future that may not be already done is from `taskManagerLocationFuture`, and is always completed in the maint hread.